Race Integrity QoL

garrettjoecox opened the pull request:
- All settings wrapped in a few `ImGui::Disable` calls based on a hidden cvar `gDisableChangingSettings`
- Added ability to prevent spoiler file generation via `gRandomizerDontGenerateSpoiler`
  - When this is enabled, an additional string is appended for seed generation to prevent the same seed from being generated without this option on.
- Support for dropping a `shipofharkinian.json` shaped file into the window, dropping it will:
  - Clear CVars
  - Apply CVars from dropped file
  - Close all unsafe windows (save editor, console, etc)
  - Clear all bindings in console
  - Print the SHA256 of the dropped file to the screen

I figure for a race the participants will be required to on stream/recording show the version of Ship of they are running, as well dragging the file in (which will print the expected SHA256 to the screen)

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh.otr.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1079667719.zip)
  - [soh-linux-compatibility.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1079667720.zip)
  - [soh-linux-performance.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1079667722.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1079667724.zip)
  - [soh-switch.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1079667725.zip)
  - [soh-wiiu.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1079667726.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1079667727.zip)
<!--- section:artifacts:end -->

garrettjoecox committed:
Add gDisableChangingSettings

Archez commented on soh/soh/Enhancements/enhancementTypes.h:
```diff
@@ -1,3 +1,5 @@
+#pragma once
```
🔥 been meaning to address this when someone found it the other day.
Although I would prefer the `#ifndef FILE_H` pattern over `#pragma once`

briaguya0 commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -925,6 +927,84 @@ bool PathTestCleanup(FILE* tfile) {
+            gui->GetGuiWindow("Collision Viewer")->Hide();
+            gui->GetGuiWindow("Save Editor")->Hide();
+            gui->GetGuiWindow("Display List Viewer")->Hide();
+            std::reinterpret_pointer_cast<LUS::ConsoleWindow>(LUS::Context::GetInstance()->GetWindow()->GetGui()->GetGuiWindow("Console"))->ClearBindings();
```
should probably use a dynamic instead of a reinterpret here

<https://stackoverflow.com/questions/332030/when-should-static-cast-dynamic-cast-const-cast-and-reinterpret-cast-be-used>

> dynamic_cast is exclusively used for handling polymorphism

`ConsoleWindow` is a `GuiWindow` so that fits

> reinterpret_cast is the most dangerous cast, and should be used very sparingly.

garrettjoecox replied:
Could you show how this can be used? I'm having trouble (just copied this from other places console window is used)

briaguya0 replied:
https://github.com/Kenix3/libultraship/blob/1b41afa9eaedb19a9073f5dd1eca2cda4dea59c8/src/controller/deviceindex/LUSDeviceIndexMappingManager.cpp#L547-L550

briaguya0 replied:
seems LUS must have a `using namespace std` hiding somewhere, here's an example with std:: in it https://github.com/HarbourMasters/Shipwright/blob/d0d1d9c4879c4f95622a9573b171a6bf13f55597/soh/soh/Enhancements/controls/SohInputEditorWindow.cpp#L294

briaguya0 commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -925,6 +927,84 @@ bool PathTestCleanup(FILE* tfile) {
+
+            uint32_t finalHash = boost::hash_32<std::string>{}(configJson.dump());
+            gui->GetGameOverlay()->TextDrawNotification(30.0f, true, "Configuration Loaded. Hash: %d", finalHash);
+        } catch (...) {
```
pokemon exception handling with no logging or anything here seems less than ideal

garrettjoecox replied:
How's that? https://github.com/HarbourMasters/Shipwright/compare/9de168c259cb91f6870870616cbfd7293d29deb5..45e5004c181c60e5e2897d5ac128c0b0017de7fa

briaguya0 replied:
beautiful

briaguya0 commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -925,6 +927,84 @@ bool PathTestCleanup(FILE* tfile) {
+
+void processDroppedFiles() {
+    const char* droppedFile = CVarGetString("gDroppedFile", "");
+    if (strcmp(droppedFile, lastDroppedFile) != 0 && strcmp(droppedFile, "") != 0) {
```
i know with rando we use `gNewFileDropped` to not need to do string compares to see if we have a new file, is there a specific reason to not use that here too?

garrettjoecox replied:
Ideally we move away from this system towards hooks or something anyway, I'm mostly just trying to make due with what we have while not touching the other parts, in hopes of not breaking something.

I mostly wasn't sure if I could even rely on `gNewFileDropped` because the file select screen sets it to 0 when it's done with it, and I'm not sure where my process is relative to that.

garrettjoecox replied:
Now using `gNewFileDropped` on the file select screen since I can guarantee the order.

briaguya0 commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -1223,6 +1303,8 @@ extern "C" void Graph_StartFrame() {
 #endif
     OTRGlobals::Instance->context->GetWindow()->StartFrame();
+
+    processDroppedFiles();
```
does this need to run always on every frame? i know with rando we only run the logic on file select which prevents any possible performance issues we might encounter by running it during gameplay

garrettjoecox replied:
Until we have a more event based approach, I'm not sure there's a better place for it, I'd rather not have to tell people to go to a specific screen for it to load their configuration

garrettjoecox replied:
Moved to every frame on the file select screen for the time being. Will circle back later on once we have a more event based approach.

briaguya0 reviewed (commented):
couple questions about the file drop implementation but overall this makes sense

garrettjoecox committed:
Add support for dropping a config file to overwrite CVars

briaguya0 commented:
looks like this needs some `#include <algorithm>`s

garrettjoecox commented:
Made a few adjustments:
- Added `gRandomizerDontGenerateSpoiler`, does what it says.
- Added some missing cvars to the reset cvars lists.
- Added a try/catch around the randomizer spoiler file parse to prevent crash on invalid file drop.
